Tunneling

cryogenian opened the pull request.

cryogenian commented:
Part of https://app.clubhouse.io/data/story/6123/ssh-tunneling-option-for-mongo-data-sources 
(The other part is FE)

cryogenian commented:
actually, I split the task it's [ch6123]

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala:
```diff
@@ -0,0 +1,134 @@
+
+  val SessionName: String = "default"
+
+  def apply[F[_]: Sync: ContextShift](config: MongoConfig): F[Disposable[F, ClusterSettings]] =
```
Do we need to use `Disposable` here? Or can we use `Stream` or `Resource`? `Disposable` is technically unsafe as the scoping is unconstrained, so we should only use it for resources which have non-lexical lifecycles.

cryogenian replied:
The tunnel should be closed when mongo-ds instance is disposed

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala:
```diff
@@ -0,0 +1,134 @@
+  final case class Tunnel(port: Int)
+
+  def openTunnel[F[_]: Sync](session: Session, address: ServerAddress): F[Tunnel] = Sync[F].delay {
+    session.connect
```
Does this need to be done on a blocking thread pool? How are the threading resources handled by JSCH?

Also, random Scala idiom: usually this would be done as `session.connect()`, since `connect` is a side-effecting nullary function. We obviously don't see this much since we simply don't side-effect, but it's nice to keep the idiom in place when we *do* have side-effects (such as here).

cryogenian replied:
JSCH creates new thread on `connect` and remove it on `disconnect`. 
It also uses `synchronized` a lot, I'll add `evalOn` to `openTunnel`, thank you!

djspiewak commented on datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala:
```diff
@@ -179,27 +187,57 @@ object MongoSpec {
+  val TunnelPassword: String = "root"
+  val TunnelPassphrase: Option[String] = Some("passphrase")
+  val TunnelHost: String = "localhost"
+  val TunnelPort: Int = 22222
```
We should consider starting to split up the tests into unit tests (which shouldn't require any of the external process dependencies, and thus don't touch SSH or mongo directly), and integration tests (which… do).

cryogenian replied:
Currently `MongoSpec` literally just checks that connection could be established and works directly with different connection strings. Other tests depend on running mongo, but don't check if it's OK, just checking aggregation, config codecs etc. What is desired behaviour here?

rintcius replied:
Could be useful to reorganize the tests a bit indeed, but maybe in another PR? 

Running the full test suite also takes quite a long time. We can then also look into making that less time consuming.

cryogenian commented:
@djspiewak 

+ Added parens after `connect` and `disconnect` 
+ `openTunnel` works on blocking pool

The tunnel is disposed when ds is disposed, so, I suppose it's OK to make it disposable and `flatMap` it with client building function, right?

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala:
```diff
@@ -0,0 +1,137 @@
+      blockingPool: BlockingContext)
+      : F[Disposable[F, ClusterSettings]] = for {
+    jsch <- mkJSch
+    sess <- mkSession(jsch, config)
```
Wondering if `mkSession` should also be done on the the `blockingPool`, it does some `synchronized` stuff here https://github.com/is/jsch/blob/e7d60f7f24dd490465eb1cf02a98cb9ce69ddfc5/src/main/java/com/jcraft/jsch/JSch.java#L169, though it's not a problem usually I guess

rintcius replied:
Not worth to use `blockingPool` just for this, but it may be good to do just everything from `mkJSch` to `tunneledSettings` in a single `evalOn`. Again, just wondering, feel free to leave as is.

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala:
```diff
@@ -0,0 +1,137 @@
+
+import scala.collection.JavaConverters._
+
+object Tunnel {
```
Maybe `Tunnel#settings` not `Settings#apply` then?

rintcius replied:
I like `Settings.apply` better because this code does not necessarily involve tunnel stuff (because of lines 47/48), it also takes care of creating the `ClusterSettings` when there is no tunneling so it's a bit broader. If the entry point would have been `viaTunnel` then `Tunnel` would have been the better name.

cryogenian commented:
:+1: for other pr for tests reorg

@rintcius 

+ Moved whole `viaTunnel` into `evalOn`, this shouldn't make things worse and definitely handle all `synchronized` 
+ Also renamed `Tunnel#apply` to `Tunnel#clusterSettings`, does it look OK? If you dislike it I'll rename `Tunnel` object into `Settings`

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala:
```diff
@@ -126,12 +126,12 @@ object Tunnel {
     result <- tunneledSettings(settings, tunnel)
-  } yield Disposable(result, ContextShift[F].evalOn(blockingPool.unwrap)(closeTunnel(sess, tunnel)))
+  } yield Disposable(result, ContextShift[F].evalOn(blockingPool.unwrap)(closeTunnel(sess, tunnel))) }
 }
```
Interesting, I would have left the `evalOn` associated with `closeTunnel` outside the scope of the first `evalOn`, so that there wouldn't be any nesting, but I think this is ok.

cryogenian replied:
AFAIU we are back in main pool after the `F[Disposable[...` is returned and closing will happen in very different context (unknown, somewhere)

rintcius replied:
Yeah that's why I think it's ok, just had to look twice :)

cryogenian commented:
@rintcius 

+ `Tunnel.clusterSettings` -> `Settings.apply`

rintcius commented:
Just this seems to be in the way:
```
[error] /home/travis/build/slamdata/quasar-datasource-mongo/datasource/src/test/scala/quasar/physical/mongo/TunnelConfigSpec.scala:21:38: Unused import
[error] import argonaut._, argonaut.Argonaut._
[error]                                      ^
[error] one error found
[error] (core / Test / compileIncremental) Compilation failed
```
